### PR TITLE
[Security] Randomize CSRF token to harden BREACH attacks

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | TODO

This PR randomize the CSRF token in each request in order to hardening the [BREACH attack](https://en.wikipedia.org/wiki/Cross-site_request_forgery)

### CODE REVIEW

**Review by nicolas-grekas — APPROVED** [review]

**Review by javiereguiluz — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
(failure is related)

**Comment by jderusse** [comment]
last failed test is about tests in  branch < `5.3` working with my patch and failed. 
The test tries to generte 2 logout links (with CSRF token) and expect there are identical which is exactly what my PR prevent.

Once this PR merge, I've to fix the test in downstream branches

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/CsrfFormLoginTest.php [hunk 1]
```diff
@@ -35,8 +35,8 @@ public function testFormLoginAndLogoutWithCsrfTokens($options)
 
         $logoutLinks = $crawler->selectLink('Log out')->links();
         $this->assertCount(2, $logoutLinks);
-        $this->assertStringContainsString('_csrf_token=', $logoutLinks[0]->getUri());
-        $this->assertSame($logoutLinks[0]->getUri(), $logoutLinks[1]->getUri());
+        $this->assertStringContainsString('_csrf_token=', $url0 = $logoutLinks[0]->getUri());
+        $this->assertSame(substr($url0, 0, strpos($url0, '_csrf_token')), substr($url1 = $logoutLinks[1]->getUri(), 0, strpos($url1, '_csrf_token')));
```
Comment (jderusse) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/CsrfFormLoginTest.php]: I think this test is useless and could be removed

**Comment by jderusse** [comment]
green \o/

**Comment by nicolas-grekas** [comment]
Thank you @jderusse.

**Comment by liweiyi88** [comment]
Hi Team, sorry to add a post-merge comment but I am learning and curious about the implementation choice used in this PR.
I am not a security expert on encryption but I read that people mentioned XOR encryption is not that secure compared with AES related algorithm. I wonder if anyone could tell if it is more secure/better to achieve encryption by using `openssl` related method like what has been implemented in [Laravel Encryption](https://github.com/illuminate/encryption/blob/master/Encrypter.php)?

**Comment by jderusse** [comment]
BREACH attack is all about guessing ~a secret~ the content of an encrypted page  by comparing the size of a compressed page.

There is nothing secret here (as an evidence: the KEY used to XOR is included in the payload).
We only need to generate random content

**Comment by stof** [comment]
To be clear, this PR does not even implement encryption at all (so there is no question about whether the encryption is secure or no)

**Comment by liweiyi88** [comment]
Cool, thanks for the explainations!

**Pull request merged**